### ISSUE TITLE
python2 test regression: (AssertionError: This is a Python 3 module)

### ISSUE THREAD

**Issue by anthraxx** [eventlet/eventlet#265]
I'm trying to package this for Arch Linux repositories, therefor I need all tests passing, python3 tests are running fine, however when testing with python2 i get following AssertionError:
    
    ======================================================================
    FAIL: Failure: AssertionError (This is a Python 3 module)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/usr/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
        addr.filename, addr.module)
      File "/usr/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
        return self.importFromDir(dir_path, fqname)
      File "/usr/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
        mod = load_module(part_fqname, fh, filename, desc)
      File "/build/python-eventlet/src/eventlet-0.17.4-py2/eventlet/green/http/__init__.py", line 2, in <module>
        assert six.PY3, 'This is a Python 3 module'
    AssertionError: This is a Python 3 module
            

From my understanding running the tests should either detect python3 and skip inclusion of that module, or expect such assertion error during test run

**Comment by temoto** [comment]
You are supposed to run tests with `tox`. Or `tox -e py27-{selects,poll,epolls}` for Python 2.7 on Linux.

**Comment by anthraxx** [comment]
I will test that, currently that seems to bring up others issues when running with python 2.7, I will see how to solve that and give feedback.

I just thought running tests directly through setup.py would be convenient (its just that single thing that fails)
https://github.com/eventlet/eventlet/blob/master/setup.py#L25

But how is python 3.5 then supposed to be tested, the tox file only contains definition for up to python 3.4. I find that kind of inflexible, whenever a new python interpreter version arrives will kill the tests if not changed accordingly and a rebuild against the new python interpreter will not be possible (as the tests fail).

In general I would be more happy to either make setup.py functional for python2 or make the tox.ini somehow more generic so a new interpreter version will just work (If its my fail just point it out i'm not that deep into tox).

**Comment by anthraxx** [comment]
so yes, all this is kind of suboptimal for me. the tox thingie does enforce to have mysql clients, postgres clients installed (errors but does not simply skip tests) and the biggest problem is version pinning. It starts downloading and building dependencies itself that are old and incompatible with the most current versions of openssl etc.

      
      OpenSSL/crypto/crl.c:6:23: error: static declaration of ‘X509_REVOKED_dup’ follows non-static declaration
       static X509_REVOKED * X509_REVOKED_dup(X509_REVOKED *orig) {
                             ^
      In file included from /usr/include/openssl/ssl.h:156:0,
                       from OpenSSL/crypto/x509.h:17,
                       from OpenSSL/crypto/crypto.h:30,
                       from OpenSSL/crypto/crl.c:3:
      /usr/include/openssl/x509.h:751:15: note: previous declaration of ‘X509_REVOKED_dup’ was here
       X509_REVOKED *X509_REVOKED_dup(X509_REVOKED *rev);
                     ^
      error: command 'gcc' failed with exit status 1
      
      ----------------------------------------
      Failed building wheel for pyopenssl
                 

Tox is nice if you want to make general development tests against multiple python iterpreter versions, however because of the last post and this post I find it inconvenient to use it (this way) for regression testing before packaging it into the distributions repositories.

For distribution packaging regression testing it would be gorgeous if I could use the setup.py nose_collected variant as it neither enforces version pinning but uses systemlibs not requires to have mysql and postgres installed (but skips those tests) and is functional for whatever major interpreter update will arrive.

I would send you 1337 imaginary internet points if we can solve that single python3 assertion test regression thingie :stuck_out_tongue_closed_eyes: 

PS: thanks for this great lib and all your work!

**Comment by temoto** [comment]
TL;DR: For your case, run `nosetests tests/` or `nosetests -sv tests/` for more helpful and verbose output.

At some point I was voting for setup.py approach because it's a widespread "API" to run `setup.py test`. But then we came to "tox is better for eventlet". Maybe `setup.py test` should just print a helpful message  explaining this all instead of doing something undetermined.

For Eventlet developer case it's important to test in as many environments and combinations of installed libraries as possible, because we're playing dangerous games with monkey patching standard and other libraries. In that tox helps a lot.

Thing with plain `nosetests` or `setup.py test` is that nose collector is too smart ass in finding test modules, but manual enumeration of each test in suite is inhumane. Now that I think of it, `__test__ = False` in that module could fix it all. Let me try it now. Nope, it didn't help, because whole module is being executed anyway, so it reaches assert line.

**Comment by anthraxx** [comment]
yes, thats exactly what i told you in my last post, i think tox is great for developer point of view. however its totally suboptimal from the distribution packaging point of view :smiley:

**Comment by anthraxx** [comment]
@temoto ok i may have missunerstood your post **nosetests2 -sv tests** seems to do the trick. I guess you can also make setup.py invoke that. For now, only the CPU consumption tests are failing so it seems to work
